File: android/app/src/main/java/com/easybc/planner/sync/shared/SingleFlightSessionValue.kt
package com.easybc.planner.sync.shared

import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * Keeps one sensitive value in process memory and coalesces concurrent loads.
 * Clearing invalidates both the cached value and any load still in flight.
 */
internal class SingleFlightSessionValue<T : Any> {
    private val mutex = Mutex()

    @Volatile
    private var cached: T? = null

    @Volatile
    private var generation: Long = 0

    suspend fun getOrLoad(loader: suspend () -> T): T {
        cached?.let { return it }
        return mutex.withLock {
            cached?.let { return@withLock it }
            val loadGeneration = generation
            val loaded = loader()
            if (generation == loadGeneration) cached = loaded
            loaded
        }
    }

    fun get(): T? = cached

    @Synchronized
    fun clear() {
        generation += 1
        cached = null
    }
}

File: web/src/sync/SingleFlightSessionValue.ts
/** Keeps one sensitive value in tab memory and coalesces concurrent loads. */
export class SingleFlightSessionValue<T> {
  private cached: T | null = null;
  private pending: Promise<T> | null = null;
  private generation = 0;

  getOrLoad(loader: () => Promise<T>): Promise<T> {
    if (this.cached !== null) return Promise.resolve(this.cached);
    if (this.pending) return this.pending;

    const loadGeneration = this.generation;
    const pending = loader()
      .then((value) => {
        if (this.generation === loadGeneration) this.cached = value;
        return value;
      })
      .finally(() => {
        if (this.pending === pending) this.pending = null;
      });
    this.pending = pending;
    return pending;
  }

  get(): T | null {
    return this.cached;
  }

  clear(): void {
    this.generation += 1;
    this.cached = null;
    this.pending = null;
  }
}
